WordPress/gutenberg
# Use consistent terminology across @wordpress/core-data

## What?
This PR unifies the language used in core-data, leaving only one term for each concept. The most prominent changes are:

* `Entity loader` replaced `kind`, `kind config`, and `entity`
* `Entity config` replaced `entity` and `kind entity` (in actions, selectors, and resolvers)
* `Entity context` replaced `entity` (in the entity provider)
* `Entity name` replaced `entity name` and `entity type` (in the entity provider)
* `state.records` replaced `state.data` 

Other than two deprecations in `selectors.js`, this PR aims to keep the public API of the core-data package the same. Every consumer of this package should continue to work in the exact same way.

## Why?
I walked through that code with @dmsnell a few days ago, and we both got super confused by all the different terms imprecisely referring sometimes to the same ideas, and sometimes to completely different ones. My intention in this PR is to make the immersion easier.

## Testing Instructions
* Comment with your opinion on whether the changes make sense
* Confirm all the tests are green – this matters less for now, let's make sure we're aligned around the idea first.

cc @gziolo @youknowriad @ellatrix @noisysocks @talldan @getdave @kevin940726 @Mamaduka
